provider/oracle/environ.go

```diff
@@ -0,0 +1,870 @@
+		apiPort = args.InstanceConfig.Controller.Config.APIPort()
+	} else {
+		// All ports are the same so pick the first.
+		apiPort = args.InstanceConfig.APIInfo.Ports()[0]
+	}
+
+	// create a new seclists
+	secLists, err := o.CreateMachineSecLists(
+		args.InstanceConfig.MachineId, apiPort)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Infof("Creating vnic sets")
+	vnicSets, err := o.ensureVnicSet(args.InstanceConfig.MachineId, tags)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// create a  new networking card used for making the instance
+	// have a public address ip
+	logger.Infof("Getting instance networks")
+	networking, err := o.getInstanceNetworks(args, secLists, []string{vnicSets.Name})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// new cloud config template based on the series
+	logger.Infof("Getting cloud config")
+	cloudcfg, err := o.getCloudInitConfig(args.InstanceConfig.Series, networking) //cloudinit.New(args.InstanceConfig.Series)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	logger.Infof("Composing userdata")
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OracleRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	logger.Debugf("oracle user data: %d bytes", len(userData))
+
+	attributes := map[string]interface{}{
+		"userdata": string(userData),
+	}
+
+	// create the instance based on the instance params
+	instance, err := o.createInstance(oci.InstanceParams{
+		Instances: []oci.Instances{
+			{
+				Shape:       spec.InstanceType.Name,
+				Imagelist:   imageName,
+				Name:        machineName,
+				Label:       args.InstanceConfig.MachineAgentServiceName,
+				Hostname:    args.InstanceConfig.MachineAgentServiceName,
+				Tags:        tags,
+				Attributes:  attributes,
+				Reverse_dns: false,
+				Networking:  networking,
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	instance.arch = &spec.Image.Arch
+	instance.instType = &spec.InstanceType
+
+	machineId := instance.machine.Name
+	timeout := 10 * time.Minute
+	if err := instance.waitForMachineStatus(ociCommon.StateRunning, timeout); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Infof("started instance %q", machineId)
+	//TODO: add config option for public IP allocation
+	logger.Infof("Associating public IP to instance %q", machineId)
+	if err := instance.associatePublicIP(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	result := &environs.StartInstanceResult{
+		Instance: instance,
+		Hardware: instance.hardwareCharacteristics(),
+	}
+
+	return result, nil
+}
+
+// StopInstances shuts down the instances with the specified IDs.
+// Unknown instance IDs are ignored, to enable idempotency.
+func (o *oracleEnviron) StopInstances(ids ...instance.Id) error {
+	//TODO: delete storage volumes
```

Comment:
It has been done. I will remove the comment. Storage is cleaned up by juju when terminating an instance.